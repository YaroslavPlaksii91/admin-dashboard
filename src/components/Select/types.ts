import { UseFormRegisterReturn, FieldErrors } from 'react-hook-form';

export type SelectProps = {
  label: string;
  name: string;
  register: UseFormRegisterReturn;
  errors: FieldErrors;
  options: {
    value: string;
    label: string;
    disabled?: boolean;
  }[];
};
