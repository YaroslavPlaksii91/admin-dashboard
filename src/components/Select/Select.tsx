import { FC } from 'react';
import { TextField, MenuItem } from '@mui/material';

import { SelectProps } from './types';

export const Select: FC<SelectProps> = ({
  label,
  name,
  register,
  errors,
  options,
  defaultValue: value,
}) => {
  const defaultValue = value ? value : options[0].value;

  return (
    <TextField
      select
      label={label}
      defaultValue={defaultValue}
      variant="outlined"
      fullWidth
      {...register}
      error={!!errors[name]}
      helperText={errors[name]?.message as string}
      InputLabelProps={{
        shrink: true,
      }}
    >
      {options.map(option => (
        <MenuItem
          key={option.value}
          value={option.value}
          disabled={option.disabled}
        >
          {option.label}
        </MenuItem>
      ))}
    </TextField>
  );
};
