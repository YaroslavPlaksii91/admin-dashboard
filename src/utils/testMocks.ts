import { UseFormRegisterReturn } from 'react-hook-form';

export const createRegister = (name: string): UseFormRegisterReturn => {
  return {
    onChange: jest.fn(),
    onBlur: jest.fn(),
    name: name,
    ref: jest.mocked,
  };
};
