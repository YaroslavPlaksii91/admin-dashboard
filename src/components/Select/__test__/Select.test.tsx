import { cleanup, render, screen } from '@testing-library/react';
import { FieldErrors, FieldError } from 'react-hook-form';

import { createRegister } from '@utils/testMocks';

import { Select } from '../Select';

describe('Select component', () => {
  afterEach(cleanup);

  const options = [
    { label: 'Option 1', value: 'option1' },
    { label: 'Option 2', value: 'option2' },
    { label: 'Option 3', value: 'option3' },
  ];
  const label = 'Select component';
  const name = 'select';
  const register = createRegister(name);
  const errors: FieldErrors = {};

  it('should be rendered', () => {
    const { asFragment } = render(
      <Select
        label={label}
        name={name}
        register={register}
        errors={errors}
        options={options}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    render(
      <Select
        label={label}
        name={name}
        register={register}
        errors={errors}
        options={options}
      />,
    );

    const labelElement = screen.getByLabelText(label);
    const selectElement = screen.getByRole('button');

    expect(labelElement).toBeInTheDocument();
    expect(selectElement).toBeInTheDocument();
  });

  it('displays error and helper text when there is an error', () => {
    const errors: FieldErrors = {
      [name]: { message: 'This field is required' } as FieldError,
    };

    render(
      <Select
        label={label}
        name={name}
        register={register}
        errors={errors}
        options={options}
      />,
    );

    const errorElement = screen.getByText('This field is required');
    expect(errorElement).toBeInTheDocument();
  });
});
