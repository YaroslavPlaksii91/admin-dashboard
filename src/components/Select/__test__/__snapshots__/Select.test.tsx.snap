// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component should be rendered 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      Select component
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby=":r0:-label :r0:"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        role="button"
        tabindex="0"
      >
        Option 1
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        name="select"
        tabindex="-1"
        value="option1"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Select component
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
